### PR TITLE
Add Migration for efficient rollback

### PR DESCRIPTION
**Pull request by kderme**
Fixes https://github.com/input-output-hk/cardano-db-sync/issues/256 and maybe https://github.com/input-output-hk/cardano-db-sync/issues/397. There are two persistent issues that needs to be fixed before merging this
- One has a pr https://github.com/yesodweb/persistent/pull/1167
- The other I haven't fixed yet. The generated migration is slightly wrong since it drops a constrant:
`    EXECUTE 'ALTER TABLE "block" DROP CONSTRAINT "block_previous_id_fkey"' ;`
without replacing it with a correct one. It looks like a persistent bug, easy to fix.

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -146,3 +146,21 @@ source-repository-package
...
+  tag: 305b11e58e6a7cff9d639d409f0ace88cf154af0
+  --sha256: 0cihcscwgs0d8wmvbm9934lj6ghdyswil26dc0nhrfzdjj7iig4g
+  subdir:
+    .
+
+source-repository-package
+  type: git
+  location: https://github.com/kderme/persistent
+  tag: 2d02c002761ba0ed05ce95aeef1dfc7bd3caa073
+  --sha256: 0jl6ph2cgrl3n2f6j7ff9gqm0scz2dwxq5vs1q1ijxlli37zfzsx
```
Comment (erikd) [inline cabal.project:162]: This is great, but I wonder if it might be more sensible to wait until these are formally released.

**Comment by kderme** [comment]
The second bug is discussed [here](https://github.com/yesodweb/persistent/issues/1169) and a fix is proposed. The fix seems to work at least for our case, so this pr is ready to test now. My manual tests show that slot rollback works fine.

**Comment by erikd** [comment]
Probably worth investigating these CI failures.

**Comment by kderme** [comment]
> Probably worth investigating these CI failures.

This seems to be related to the --only-dependencies flag of ci and the fact that I use a custom branch of persistent at the cabal.project. Eventually though we won't have a custom branch (unless we want this pr merged asap).

**Comment by erikd** [comment]
Rebased and squashed down this PR. Hope it passes CI.

**Pull request merged**